src/node_worker.cc

```diff
@@ -346,7 +346,7 @@ void Worker::Run() {
           more = uv_loop_alive(&data.loop_);
           if (more && !is_stopped()) continue;
 
-          EmitBeforeExit(env_.get());
+          if (!isolate_->IsExecutionTerminating()) EmitBeforeExit(env_.get());
```

Comment:
This is not a good bug fix, imo:

- It only fixes this for Worker threads, but the bug itself applies to the main thread as well.
- It’s inherently racy, because there can always be a `TerminateExecution()` call after the `IsExecutionTerminating()` check, including during the call into JS itself. The only proper way to fix this is to handle the empty `MaybeLocal<>` in `EmitBeforeExit` itself. 
  - I don’t think we can do anything besides ignore it at this point, and then make `EmitBeforeExit` return a `Maybe<>` itself as an ABI-breaking change.

Reply:
I thought my one-liner was rather elegant because it neatly sidesteps the rabbit hole that is `EmitBeforeExit()` calling `Environment::RunBeforeExitCallbacks()` calling `AsyncWrap::DestroyAsyncIdsCallback()` calling arbitrary JS code.

But okay, since you asked for it - updated to be more rigorous, PTAL.